File: kotlin/tutorials/tour-of-orchestration-kotlin/src/main/kotlin/my/example/communication/service.kt
import dev.restate.sdk.Context
import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Service
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.endpoint
import java.math.BigDecimal
import java.time.Duration
import java.time.Instant
import java.time.ZonedDateTime
import java.util.*

data class PurchaseTicketRequest(
    val ticketId: String,
    val concertDateTime: String,
    val price: BigDecimal,
    val customerEmail: String
)

// dayBefore calculates delay until day before concert
fun dayBefore(concertDate: String): Duration {
    return try {
        val concertDateTime = ZonedDateTime.parse(concertDate)
        val concertTimeMillis = concertDateTime.toInstant().toEpochMilli()
        val now = Instant.now().toEpochMilli()
        val oneDayMillis = 24 * 60 * 60 * 1000
        val delay = concertTimeMillis - now - oneDayMillis
        
        if (delay < 0) {
            println("Reminder date is in the past, cannot schedule reminder.")
            Duration.ofMillis(0)
        } else {
            println("Scheduling reminder for $concertDate with delay ${delay}ms")
            Duration.ofMillis(delay)
        }
    } catch (e: Exception) {
        println("Invalid date format: $concertDate")
        Duration.ofMillis(0)
    }
}

@Service
class PaymentService {
    @Handler
    suspend fun charge(ctx: Context, req: PurchaseTicketRequest): String {
        // Simulate payment processing
        val paymentId = ctx.random().nextUUID().toString()
        println("Processing payment for ticket ${req.ticketId} with payment ID $paymentId")
        return paymentId
    }
}

@Service
class EmailService {
    @Handler
    suspend fun emailTicket(ctx: Context, req: PurchaseTicketRequest) {
        println("Sending ticket to ${req.customerEmail} for concert on ${req.concertDateTime}")
    }

    @Handler
    suspend fun sendReminder(ctx: Context, req: PurchaseTicketRequest) {
        println("Sending reminder for concert on ${req.concertDateTime} to ${req.customerEmail}")
    }
}

@Service
class ConcertTicketingService {
    @Handler
    suspend fun buy(ctx: Context, req: PurchaseTicketRequest): String {
        // Request-response call - wait for payment to complete
        val payRef = PaymentServiceClient.fromContext(ctx).charge(req).await()

        // One-way message - fire and forget ticket delivery
        EmailServiceClient.fromContext(ctx).send().emailTicket(req)

        // Delayed message - schedule reminder for day before concert
        val delay = dayBefore(req.concertDateTime)
        EmailServiceClient.fromContext(ctx).send().sendReminder(req, delay)

        return "Ticket purchased successfully with payment reference: $payRef"
    }
}

fun main() {
    RestateHttpServer.listen(endpoint {
        bind(PaymentService())
        bind(EmailService())
        bind(ConcertTicketingService())
    })
}

File: go/tutorials/tour-of-orchestration-go/communication/service.go
package communication

import (
	"fmt"
	"github.com/restatedev/sdk-go"

	"github.com/restatedev/examples/go/tutorials/tour-of-orchestration-go/shared"
)

type PaymentService struct{}

func (PaymentService) Charge(ctx restate.Context, req shared.PurchaseTicketRequest) (string, error) {
	// Simulate payment processing
	paymentId := restate.Rand(ctx).UUID().String()
	fmt.Printf("Processing payment for ticket %s with payment ID %s\n", req.TicketId, paymentId)
	return paymentId, nil
}

type EmailService struct{}

func (EmailService) EmailTicket(ctx restate.Context, req shared.PurchaseTicketRequest) error {
	fmt.Printf("Sending ticket to %s for concert on %s\n", req.CustomerEmail, req.ConcertDate)
	return nil
}

func (EmailService) SendReminder(ctx restate.Context, req shared.PurchaseTicketRequest) error {
	fmt.Printf("Sending reminder for concert on %s to %s\n", req.ConcertDate, req.CustomerEmail)
	return nil
}

type ConcertTicketingService struct{}

func (ConcertTicketingService) Buy(ctx restate.Context, req shared.PurchaseTicketRequest) (string, error) {
	// Request-response call - wait for payment to complete
	payRef, err := restate.Service[string](ctx, "PaymentService", "Charge").Request(req)
	if err != nil {
		return "", err
	}

	// One-way message - fire and forget ticket delivery
	restate.Service[restate.Void](ctx, "EmailService", "EmailTicket").Send(req)

	// Delayed message - schedule reminder for day before concert
	delay := shared.DayBefore(req.ConcertDate)
	restate.Service[restate.Void](ctx, "EmailService", "SendReminder").
		Send(req, restate.WithDelay(delay))

	return fmt.Sprintf("Ticket purchased successfully with payment reference: %s", payRef), nil
}
